src/FSharp.Analyzers/ImmutableCollectionEqualityAnalyzer.fs

```diff
@@ -0,0 +1,93 @@
+module GR.FSharp.Analyzers.ImmutableCollectionEqualityAnalyzer
+
+open FSharp.Analyzers.SDK
+open FSharp.Analyzers.SDK.TASTCollecting
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Text
+open FSharp.Compiler.Symbols.FSharpExprPatterns
+
+
+[<Literal>]
+let Code = "GRA-IMMUTABLECOLLECTIONEQUALITY-001"
+
+let isImmutableType (basicQualifiedName : string) (e : FSharpExpr) =
+    let e =
+        match e with
+        | Coerce (_, e)
+        | e -> e
+
+    let t = e.Type
+
+    t.BasicQualifiedName = basicQualifiedName
+    && t.TypeDefinition.Assembly.SimpleName = "System.Collections.Immutable"
+
+let isImmutableDictionaryType (e : FSharpExpr) =
+    isImmutableType "System.Collections.Immutable.ImmutableDictionary`2" e
+
+let isImmutableHashSetType (e : FSharpExpr) =
+    isImmutableType "System.Collections.Immutable.ImmutableHashSet`1" e
+
+let mkMessage typeName m =
+    {
+        Type = "ImmutableCollectionEqualityAnalyzer"
+        Message =
+            $"%s{typeName} does not implement structural equality. Use Object.ReferenceEquals instead .Equals, = or <> to emphasize this."
```

Comment:
```suggestion
            $"%s{typeName} does not implement structural equality. Use Object.ReferenceEquals instead of .Equals, = or <> to emphasize this."
```